### PR TITLE
[issue 165] javascript expression at definition URL

### PR DESCRIPTION
**Pull request by Ginxo**
**Thank you for submitting this pull request**

please do not check index.js since the file is automatically generated.

fix #165

Not sure about `%{}` characters to identify a javascript expression. We can't use `${}` since it can be part of the javascript expression. Any better idea?

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

- https://github.com/kiegroup/build-chain-configuration-reader/pull/23
- https://github.com/kiegroup/github-action-build-chain/pull/170

### CODE REVIEW

File: test/lib/util/reader-util.test.js [hunk 1]
```diff
@@ -1,5 +1,56 @@
...
+  // Act
+  const result = executeUrlExpressions(string);
+
+  // Assert
+  expect(result).toBe(
+    "https://raw.githubusercontent.com/kiegroup/droolsjbpm-build-bootstrap/8.2.3/.ci/pull-request-config.yml"
+  );
+});
+
+test("executeUrlExpressions with 2 expressions", () => {
```
Comment (radtriste) [inline test/lib/util/reader-util.test.js:38]: Could you also test with an expression and no matching of version ?

for example with
```
process.env = Object.assign(process.env, {
    GITHUB_BASE_REF: "1.2.3"
  });
```
and
```
const string =
    "https://raw.githubusercontent.com/kiegroup/droolsjbpm-build-bootstrap/%{process.env.GITHUB_BASE_REF.replace(/(\\d*)\\.(.*)\\.(.*)/g, (m, n1, n2, n3) => `${+n1+7}.${n2}.${n3}`)}/.ci/pull-request-config.yml";
```
This should give "https://raw.githubusercontent.com/kiegroup/droolsjbpm-build-bootstrap/main/.ci/pull-request-config.yml"

**Comment by radtriste** [comment]
I guess `%{}` should be ok

File: README.md [hunk 2]
```diff
@@ -234,8 +237,14 @@ it will produce
 it can be expressed like a simple url (http(s)://whateverurl.domain/whateverfile.txt) or it can be expressed using place holders like `https://raw.githubusercontent.com/${GROUP}/${PROJECT}/${BRANCH}/dependencies.yaml` to get `dependencies.yaml` file from source group, project and branch tokens (no need to use all of them at the same time).
 
 **Place holders**
+
 The place holders must be wrapped between `${}` expression and you can use any place holder you want.
 
+**Javascript expressions**
+
+It is possible to define javascript expressions wrapped by `%{}`. So for instance a definition file string like
+`` https://raw.githubusercontent.com/kiegroup/droolsjbpm-build-bootstrap/%{process.env.GITHUB_BASE_REF.replace(/(\\d*)\\.(.*)\\.(.*)/g, (m, n1, n2, n3) => `${+n1+7}.${n2}.${n3}`)}/.ci/pull-request-config.yml `` in case `GITHUB_BASE_REF` environment variable's value is `1.2.x`, the expression between `%{}` will be evaluated and final definition file will be `https://raw.githubusercontent.com/kiegroup/droolsjbpm-build-bootstrap/8.2.x/.ci/pull-request-config.yml`.
```
Comment (rsynek) [inline README.md:246]: Just out of curiosity, why not enabling the expressions with the `${}` syntax?

File: test/lib/util/reader-util.test.js (hunk 1 above)
Comment (Ginxo) [reply test/lib/util/reader-util.test.js:38]: @radtriste good catch. I guess this is the test you mean https://github.com/kiegroup/build-chain-configuration-reader/pull/23/commits/63acef64c83fba71d82fc2c4e0ea5424a4294aec

Comment (radtriste) [reply test/lib/util/reader-util.test.js:38]: yes

File: README.md (hunk 2 above)
Comment (Ginxo) [reply README.md:246]: @rsynek I would prefer to use `${}` as you suggest but, let's suppose a `definition-file` url like ```"https://raw.githubusercontent.com/kiegroup/droolsjbpm-build-bootstrap/${process.env.GITHUB_BASE_REF.replace(/(\\d*)\\.(.*)\\.(.*)/g, (m, n1, n2, n3) => `${+n1+7}.${n2}.${n3}`)}/.ci/pull-request-config.yml"``` the case `${whatever${whatever}}` is found. I can't find the regex to get `${whatever${whatever}}`.
The regex `/\${([^}]+)}/g` will return:

- ${process.env.GITHUB_BASE_REF.replace(/(\\d*)\\.(.*)\\.(.*)/g, (m, n1, n2, n3) => `${+n1+7}
- ${n2}
- ${n3}

File: test/lib/util/reader-util.test.js (hunk 1 above)
Comment (Ginxo) [reply test/lib/util/reader-util.test.js:38]: thanks man!

File: README.md (hunk 2 above)
Comment (Ginxo) [reply README.md:246]: The other option is to use `${{ }}` and regex `\${{(.*)}}`

**Pull request merged**